src/contexts/globalLayout/GlobalLayoutContext.tsx

```diff
@@ -0,0 +1,371 @@
+
+const GlobalLayoutContextProvider = memo(({ children }: Props) => {
+  const query = useLazyLoadQuery<GlobalLayoutContextQuery>(
+    graphql`
+      query GlobalLayoutContextQuery {
+        ...GlobalLayoutContextNavbarFragment
+      }
+    `,
+    {}
+  );
+
+  // whether the global banner is visible
+  const [isBannerVisible, setBannerVisible] = useState(false);
+
+  /**
+   * the action that triggered the fade animation.
+   *
+   * there are 3 types of triggers:
+   * - route-based setting determined by `GalleryRoute.tsx`
+   * - scroll-based setting determined by scroll position
+   * - hover-based setting determined by mouse position
+   */
+  const [fadeType, setFadeType] = useState<FadeTriggerType>('route');
+
+  /**
+   * we track whether the child route is in suspense in both React State and Ref:
+   * - `state` is used to determine whether we should debounce or throttle nav animation
+   * - `ref` is used to stabilizing callbacks that areused for scroll, hover handling
+   *
+   * why do we care if the app is in suspense?
+   * - it allows us to disable visibility of the navbar against the loading fallback
+   *
+   * can't we just address the above with z-index?
+   * - no, because when both the loading fallback and the navbar have between 0-1 opacity,
+   *   both elements are displayed on the screen.
+   */
+  const [isPageInSuspenseState, setIsPageInSuspenseState] = useState(false);
+  const isPageInSuspenseRef = useRef(false);
+
+  /**
+   * navbar-related states.
+   *
+   * - `isNavbarEnabled` is the raw state that's set by "dumb" components / interactions
+   *   that simply wanna display / hide the nav, according to its internal logic
+   * - we then need to `debounce` or `throttle` this state because many interactions could be
+   *   fighting to set this state
+   * - the difference between debouncing vs. throttling comes down to whether we need to
+   *   delay the navbar fade animation vs. display it immediately, respectively.
+   * - `wasNavbarVisible` is used to determine whether we should continue displaying the navbar
+   *   across transitions. for example, if the user navigates from /a => /b and the navbar
+   *   is visible in both scenarios, we should keep it in view, without any fade animations.
+   */
+  const [isNavbarEnabled, setIsNavbarEnabled] = useState(false);
+  const debounced = useDebounce(isNavbarEnabled, FADE_TRANSITION_TIME_MS);
+  const throttled = useThrottle(isNavbarEnabled, FADE_TRANSITION_TIME_MS);
+  const isNavbarVisible = useMemo(() => {
+    if (isPageInSuspenseState && isNavbarEnabled) {
+      return debounced;
+    }
+    return throttled;
+  }, [isNavbarEnabled, isPageInSuspenseState, throttled, debounced]);
+  const wasNavbarVisible = usePrevious(isNavbarVisible) ?? false;
+
+  // whether the route wants the navbar disabled. this is an override where scrolling / hovering
+  // will not make the navbar appear
+  const forcedHiddenByRouteRef = useRef(false);
```

Comment:
how come these are refs? I see the comments on L96-98 but I don't fully get it 🤓

I'm guessing we want to update some values without causing a re-render or other effects? but unsure what happens if we don't use ref

Reply:
so i've been leaning more towards using `ref` in these complex scenarios to avoid side-effects.

**TL;DR – use `state` only when you need to**

let's flip the question: when would you use `state`? state is useful when you want to:
1. trigger a re-render if the value changes (e.g. button disabled => enabled after X change)
2. trigger a side-effect if the value changes (e.g. begin fetching data after X change)

so using `state` is very deliberate and comes with the intention of updating your react view, side effects.

when would you use a `ref`? when you explicitly _don't_ want to trigger re-renders, but still depend on that value when evaluating something.

knowing this, let's think about the purpose of `forcedHiddenByRouteRef`. what does this piece of information do? it tells us whether the route should be completely invisible, despite other factors that try to make it visible.

let's see where `forcedHiddenByRouteRef` is used:
<img width="505" alt="image" src="https://user-images.githubusercontent.com/12162433/168905409-5aee2dde-0258-45b3-923b-001630db59d8.png">
<img width="622" alt="image" src="https://user-images.githubusercontent.com/12162433/168905490-f7d1e06a-ef7f-45fa-8107-0b3b12e0ee5b.png">

it's used in these callbacks. when a callback is fired, e.g. `handleFadeNavbarOnScroll`, it looks up whether the navbar should be hidden all the time. if so, it opts out of displaying the navbar, even if the user scrolled to a position where it should be visible.

why is this not in `state`? because it doesn't have to be!
1. the value is being looked up whenever the callback is fired
2. the value is NOT being utilized in the render, or useEffect

in fact, if it _was_ in state, it would de-stabilize our callback and potentially cause unwanted side-effects.

```js
// if we use state
const handleScroll = useCallback(() => {
  if (forcedHiddenByRoute) { ... }
}, [forcedHiddenByRoute]) <-- a new callback would be defined
                              every time this value changes,
                              which is bad if we're putting this
                              callback inside of a document event
                              listener

// if we use ref
const handleScroll = useCallback(() => {
  if (forcedHiddenByRouteRef.current) { ... }
}, []) <-- the callback can remain pure, which is critical
           for event listeners
```

Reply:
ok here's a commit on another PR that's directly relevant! using `state` caused a bug / race condition so i had to switch to using ref: https://github.com/gallery-so/gallery/pull/755/commits/f813a13aa6148443da3722b14b6fef64910e5859